File: 78_Subsets_Recursion.py
# Time Complexity : O(2^n)[n = number of elements in the list]
# Space Complexity : O(2^n * n) [n = number of element in the list]
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Problem Approach
# 1. At each recusive call we have the option to either choose the element at current index or skip it
# 2. A new list is created for each recusive call
# 3. The recusion terminates when the index reaches the length of the input list
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        def helper(nums, index, curr):
            nonlocal res
            ## base
            if index == len(nums):
                res.append(curr)
                return
            
            ## body
            
            # don't choose
            helper(nums, index+1, curr)
            
            # choose
            new_list = [num for num in curr]
            new_list.append(nums[index])
            helper(nums, index+1, new_list)
        helper(nums, 0, [])
        return res

File: 78-SubsetsRecursion.java
/**
 * 
 * Time Complexity : 
 * Space Complexity : 
 * Did this code successfully run on Leetcode : Yes
 * Any problem you faced while coding this :
 * 
 *
 
 */
class Solution {
    
    List<List<Integer>> result;
    public List<List<Integer>> subsets(int[] nums) {
        
        result = new ArrayList<>();
        
        if(nums == null || nums.length == 0) return result;
        
        helper(nums, new ArrayList<>(), 0);
        return result;
    }
    
    private void helper(int [] nums, List<Integer> temp, int index){
        
        //base
        if(index == nums.length) {
            result.add(temp);
            return;
        }

        
        //logic
        
        //dont choose
        helper(nums, new ArrayList<>(temp), index + 1);
        
        //choose
        temp.add(nums[index]);
        helper(nums, new ArrayList<>(temp), index + 1);
        
    }
}